lib/MongoDB/ChangeStream.pm

```diff
@@ -27,80 +27,110 @@ use MongoDB::Cursor;
 use MongoDB::Op::_ChangeStream;
 use MongoDB::Error;
 use Safe::Isa;
+use BSON::Timestamp;
 use MongoDB::_Types qw(
     MongoDBCollection
     ArrayOfHashRef
+    Boolish
+    Intish
+    BSONTimestamp
+    ClientSession
 );
 use Types::Standard qw(
-    Bool
     InstanceOf
-    Int
     HashRef
     Maybe
     Str
+    Num
 );
 
 use namespace::clean -except => 'meta';
 
 has _result => (
     is => 'rw',
     isa => InstanceOf['MongoDB::QueryResult'],
+    init_arg => undef,
     lazy => 1,
     builder => '_build_result',
     clearer => '_clear_result',
 );
 
-has client => (
+has _client => (
     is => 'ro',
     isa => InstanceOf['MongoDB::MongoClient'],
+    init_arg => 'client',
     required => 1,
 );
 
 has _op_args => (
     is => 'ro',
     isa => HashRef,
-    default => sub { {} },
     init_arg => 'op_args',
+    required => 1,
 );
 
-has collection => (
-    is => 'ro',
-    isa => MongoDBCollection,
-);
-
-has pipeline => (
+has _pipeline => (
     is => 'ro',
     isa => ArrayOfHashRef,
+    init_arg => 'pipeline',
     required => 1,
 );
 
-has full_document => (
+has _full_document => (
     is => 'ro',
     isa => Str,
+    init_arg => 'full_document',
     predicate => '_has_full_document',
 );
 
 has _resume_token => (
     is => 'rw',
     init_arg => 'resume_after',
     predicate => '_has_resume_token',
-    lazy => 1,
 );
 
-has all_changes_for_cluster => (
+has _all_changes_for_cluster => (
     is => 'ro',
-    isa => Bool,
+    isa => Boolish,
+    init_arg => 'all_changes_for_cluster',
     default => sub { 0 },
 );
 
 has _changes_received => (
     is => 'rw',
+    isa => Boolish,
+    init_arg => 'changes_received',
     default => sub { 0 },
 );
 
-has start_at_operation_time => (
+has _start_at_operation_time => (
+    is => 'ro',
+    isa => BSONTimestamp,
+    init_arg => 'start_at_operation_time',
+    predicate => '_has_start_at_operation_time',
+    coerce => sub {
+        ref($_[0]) ? $_[0] : BSON::Timestamp->new($_[0])
```

Comment:
If you're wanting to allow seconds, I'd prefer to have this be `BSON::Timestamp->new( seconds => $_[0])` rather than relying on BUILDARGS to fix up the legacy constructor format.